.github/workflows/main.yml

```diff
@@ -192,5 +192,24 @@ jobs:
         uses: ./.github/actions/setup
         with:
           node: ${{ env.NODE_VERSION }}
-      - name: Lint Rust client
-        run: pnpm clients:rust:lint
+      - name: Set environment variables
+        shell: bash
+        run: pnpm zx ./scripts/ci/set-variables.mjs
+      - name: Install Rustfmt
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ env.RUSTFMT_TOOLCHAIN }}
+          components: rustfmt
+      - name: Install Clippy
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ env.CLIPPY_TOOLCHAIN }}
+          components: clippy
+      - name: Programs
+        run: |
+          pnpm programs:format --check
+          pnpm programs:clippy
```

Comment:
Sorry, maybe my comment wasn't clear, but can you move the program format / clippy checks to earlier in the pipeline? It could be something like:

```
lint_programs:
    name: Lint Programs
    runs-on: ubuntu-latest
    steps:
      - name: Git checkout
        uses: actions/checkout@v4
      - name: Setup environment
        uses: ./.github/actions/setup
        with:
          node: ${{ env.NODE_VERSION }}
      - name: Set environment variables
        shell: bash
        run: pnpm zx ./scripts/ci/set-variables.mjs
      - name: Install Rustfmt
        uses: dtolnay/rust-toolchain@master
        with:
          toolchain: ${{ env.RUSTFMT_TOOLCHAIN }}
          components: rustfmt
      - name: Install Clippy
        uses: dtolnay/rust-toolchain@master
        with:
          toolchain: ${{ env.CLIPPY_TOOLCHAIN }}
          components: clippy
      - name: Programs
        run: |
          pnpm programs:format --check
          pnpm programs:clippy
```

Even better, you could split them into two separate steps:

```
fmt_programs:
    name: Format Programs
    runs-on: ubuntu-latest
    steps:
      - name: Git checkout
        uses: actions/checkout@v4
      - name: Setup environment
        uses: ./.github/actions/setup
        with:
          node: ${{ env.NODE_VERSION }}
      - name: Set environment variables
        shell: bash
        run: pnpm zx ./scripts/ci/set-variables.mjs
      - name: Install Rustfmt
        uses: dtolnay/rust-toolchain@master
        with:
          toolchain: ${{ env.RUSTFMT_TOOLCHAIN }}
          components: rustfmt
      - name: Programs
        run: |
          pnpm programs:format --check
clippy_programs:
    name: Clippy Programs
    runs-on: ubuntu-latest
    steps:
      - name: Git checkout
        uses: actions/checkout@v4
      - name: Setup environment
        uses: ./.github/actions/setup
        with:
          node: ${{ env.NODE_VERSION }}
      - name: Set environment variables
        shell: bash
        run: pnpm zx ./scripts/ci/set-variables.mjs
      - name: Install Clippy
        uses: dtolnay/rust-toolchain@master
        with:
          toolchain: ${{ env.CLIPPY_TOOLCHAIN }}
          components: clippy
      - name: Programs
        run: pnpm programs:clippy
```

Reply:
I am working on moving all linting steps to the beginning, including JS, programs and rust client.